fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -280,6 +292,83 @@ public FDBLuceneFileReference getFDBLuceneFileReference(@Nonnull final String na
         return context.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_GET_FILE_REFERENCE, getFDBLuceneFileReferenceAsync(name));
     }
 
+    public void setFieldInfoId(final String filename, final long id, final ByteString bitSet) {
+        final FDBLuceneFileReference reference = context.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_GET_FILE_REFERENCE,
+                getFDBLuceneFileReferenceAsync(filename));
+        if (reference == null) {
+            throw new RecordCoreException("Reference not found")
+                    .addLogInfo(LuceneLogMessageKeys.FILE_NAME, filename);
+        }
+        reference.setFieldInfosId(id);
+        reference.setFieldInfosBitSet(bitSet);
+        writeFDBLuceneFileReference(filename, reference);
+    }
+
+    public long writeFieldInfo(byte[] value) {
+        long id;
+        if (Boolean.TRUE.equals(context.asyncToSync(
+                LuceneEvents.Waits.WAIT_LUCENE_READ_FIELD_INFOS,
+                getAllFieldInfos().thenApply(Map::isEmpty)))) {
+            id = GLOBAL_FIELD_INFOS_ID;
+        } else {
+            id = getIncrement();
+        }
+        return writeFieldInfo(value, id);
+    }
+
+    public long writeFieldInfo(byte[] value, long id) {
+        if (id == 0) {
+            throw new RecordCoreArgumentException("FieldInfo id should never be 0");
+        }
+        byte[] key = fieldInfosSubspace.pack(id);
+        context.increment(LuceneEvents.Counts.LUCENE_WRITE_SIZE, key.length + value.length);
+        context.increment(LuceneEvents.Counts.LUCENE_WRITE_CALL);
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace(getLogMessage("Write lucene stored field infos data",
+                    LuceneLogMessageKeys.DATA_SIZE, value.length,
+                    LuceneLogMessageKeys.ENCODED_DATA_SIZE, value.length));
+        }
+        context.ensureActive().set(key, value);
+        // Add the entry in the cached map from id->bytes, loading the cache if it hasn't been loaded yet.
+        context.asyncToSync(
+                LuceneEvents.Waits.WAIT_LUCENE_READ_FIELD_INFOS,
+                getAllFieldInfos().thenApply(map -> map.put(id, value)));
+        return id;
+    }
+
+    public byte[] readFieldInfo(long id) {
+        return context.asyncToSync(
+                LuceneEvents.Waits.WAIT_LUCENE_READ_FIELD_INFOS,
+                getAllFieldInfos().thenApply(allFieldInfos -> allFieldInfos.get(id)));
+    }
+
+    public byte[] readGlobalFieldInfos() {
+        return context.asyncToSync(
+                LuceneEvents.Waits.WAIT_LUCENE_READ_FIELD_INFOS,
+                getAllFieldInfos().thenApply(allFieldInfos -> allFieldInfos.get(GLOBAL_FIELD_INFOS_ID)));
+    }
+
+    public long updateGlobalFieldInfos(final byte[] fieldInfos) {
+        return writeFieldInfo(fieldInfos, GLOBAL_FIELD_INFOS_ID);
+    }
+
+    @VisibleForTesting
+    public CompletableFuture<Map<Long, byte[]>> getAllFieldInfos() {
```

Comment:
Every call to this appears to immediately do a `context.asyncToSync`, so perhaps this ought to do it itself. Or this could be `getAllFieldInfosAsync` and a new method could combine with that.

Reply:
changed `getAllFieldInfos` to call `asyncToSync`